x/protorev/keeper/rebalance.go

```diff
@@ -0,0 +1,160 @@
+	// Get highest liquidity pool ID for the input coin and uosmo
+	conversionPoolID, err := k.GetOsmoPool(ctx, inputCoin.Denom)
+	if err != nil {
+		return profit, err
+	}
+
+	// Get the pool
+	conversionPool, err := k.gammKeeper.GetPoolAndPoke(ctx, conversionPoolID)
+	if err != nil {
+		return profit, err
+	}
+
+	// Calculate the amount of uosmo that we can get if we swapped the
+	// profited amount of the orignal asset through the highest uosmo liquidity pool
+	conversionTokenOut, err := conversionPool.CalcOutAmtGivenIn(ctx, sdk.NewCoins(sdk.NewCoin(inputCoin.Denom, profit)), types.OsmosisDenomination, conversionPool.GetSwapFee(ctx))
+	if err != nil {
+		return profit, err
+	}
+
+	// return the profit denominated in uosmo
+	return conversionTokenOut.Amount, nil
+}
+
+// EstimateMultihopProfit estimates the profit for a given route
+// by estimating the amount out given the amount in for the first pool in the route
+// and then subtracting the amount in from the amount out to get the profit
+func (k Keeper) EstimateMultihopProfit(ctx sdk.Context, inputDenom string, amount sdk.Int, route gammtypes.SwapAmountInRoutes) (sdk.Coin, sdk.Int, error) {
+	tokenIn := sdk.NewCoin(inputDenom, amount)
+	amtOut, err := k.gammKeeper.MultihopEstimateOutGivenExactAmountIn(ctx, route, tokenIn)
+	if err != nil {
+		return sdk.NewCoin(types.OsmosisDenomination, sdk.ZeroInt()), sdk.ZeroInt(), err
+	}
+	profit := amtOut.Sub(tokenIn.Amount)
+	return tokenIn, profit, nil
+}
+
+// FindMaxProfitRoute runs a binary search to find the max profit for a given route
+func (k Keeper) FindMaxProfitForRoute(ctx sdk.Context, route gammtypes.SwapAmountInRoutes, inputDenom string) (sdk.Coin, sdk.Int, error) {
+	left := 0
+	right := len(types.InputAmountList) - 1
+	midPosition := 0
+	iteration := 0
+
+	for left < right {
+		midPosition = (left + right) / 2
+
+		// Short circuit profit searching if there is an error in the GAMM module
+		_, profitMidPosition, err := k.EstimateMultihopProfit(ctx, inputDenom, types.InputAmountList[midPosition], route)
+		if err != nil {
+			return sdk.Coin{Denom: types.OsmosisDenomination, Amount: sdk.ZeroInt()}, sdk.ZeroInt(), err
+		}
+
+		// Short circuit profit searching if there is an error in the GAMM module
+		_, profitMidPositionPlusOne, err := k.EstimateMultihopProfit(ctx, inputDenom, types.InputAmountList[midPosition+1], route)
+		if err != nil {
+			return sdk.Coin{Denom: types.OsmosisDenomination, Amount: sdk.ZeroInt()}, sdk.ZeroInt(), err
+		}
+
+		// Reduce subspace to search for max profit
+		if profitMidPosition.LTE(profitMidPositionPlusOne) {
+			left = midPosition + 1
+		} else if profitMidPosition.GT(profitMidPositionPlusOne) {
+			right = midPosition
+		}
+
+		// Circuit breaker to prevent unbounded runtime
+		iteration += 1
+		if iteration >= 20 {
+			break
+		}
+	}
+
+	tokenIn, profit, err := k.EstimateMultihopProfit(ctx, inputDenom, types.InputAmountList[left], route)
```

Comment:
This looks like some wasted computation

I think in all cases except when the iteration just stops because we've iterated 20 times already, this is just recalculating the profit that you've already calculated for midPositionPlusOne or midPosition (depending on whether profitMidPosition <= profitMidPositionPlusOne)

Reply:
Yeah let me make check a conditional here instead

Reply:
Changed to set them with the profitMidPosition <= profitMidPositionPlusOne check